java-jacoco/src/main/java/org/sonar/plugins/jacoco/AbstractAnalyzer.java

```diff
@@ -182,9 +177,62 @@ public final void readExecutionData(File jacocoExecutionData, SensorContext cont
     }
   }
 
+  public static boolean isCurrentReportFormat(File jacocoExecutionData) throws IOException {
+    DataInputStream dis = null;
+    try {
+      dis = new DataInputStream(new FileInputStream(jacocoExecutionData));
+      byte firstByte = dis.readByte();
+      Preconditions.checkState(firstByte == ExecutionDataWriter.BLOCK_HEADER);
+      Preconditions.checkState(dis.readChar() == ExecutionDataWriter.MAGIC_NUMBER);
+      char version = dis.readChar();
+      return version == ExecutionDataWriter.FORMAT_VERSION;
+    } finally {
+      if (dis != null) {
+        dis.close();
+      }
+    }
+  }
+
+  private boolean readJacocoReport(File jacocoExecutionData, ExecutionDataVisitor executionDataVisitor) throws IOException {
+    return readJacocoReport(jacocoExecutionData, executionDataVisitor, executionDataVisitor);
+  }
+
+  /**
+   * Read JaCoCo report determining the format to be used.
+   * @param jacocoExecutionData binary report file.
+   * @param executionDataVisitor visitor to store execution data.
+   * @param sessionInfoStore visitor to store info session.
+   * @return true if binary format is the latest one.
+   * @throws IOException in case of error or binary format not supported.
+   */
+  public static boolean readJacocoReport(File jacocoExecutionData, IExecutionDataVisitor executionDataVisitor, ISessionInfoVisitor sessionInfoStore) throws IOException {
+    JaCoCoExtensions.LOG.info("Analysing {}", jacocoExecutionData);
+    InputStream inputStream = null;
+    boolean currentReportFormat = true;
+    try {
+      inputStream = new BufferedInputStream(new FileInputStream(jacocoExecutionData));
+      currentReportFormat = isCurrentReportFormat(jacocoExecutionData);
```

Comment:
@benzonico IMO possible to use just one stream instead of opening file multiple times - see methods ``mark`` and ``reset`` in BufferedInputStream, but this might be considered as minor optimization